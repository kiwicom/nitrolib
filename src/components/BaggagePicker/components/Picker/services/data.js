// @flow
const data = [
  {
    originalIndex: 5,
    bagType: "holdBag",
    price: {
      currency: "EUR",
      amount: 0,
    },
    items: {},
  },
  {
    originalIndex: 6,
    bagType: "holdBag",
    price: {
      currency: "EUR",
      amount: 10,
    },
    items: {
      "0": {
        amount: 1,
        category: "holdBag",
        restrictions: {
          weight: 10,
          height: 52,
          width: 26,
          length: 78,
          dimensions_sum: 156,
        },
      },
    },
  },
  {
    originalIndex: 7,
    bagType: "holdBag",
    price: {
      currency: "EUR",
      amount: 25,
    },
    items: {
      "1": {
        amount: 1,
        category: "holdBag",
        restrictions: {
          weight: 15,
          height: 52,
          width: 26,
          length: 78,
          dimensions_sum: 156,
        },
      },
    },
  },
  {
    originalIndex: 8,
    bagType: "holdBag",
    price: {
      currency: "EUR",
      amount: 30,
    },
    items: {
      "2": {
        amount: 1,
        category: "holdBag",
        restrictions: {
          weight: 25,
          height: 52,
          width: 26,
          length: 78,
          dimensions_sum: 156,
        },
        conditions: {
          is_priority: ["FR", "W6"],
        },
      },
    },
  },
  {
    originalIndex: 9,
    bagType: "holdBag",
    price: {
      currency: "EUR",
      amount: 50,
    },
    items: {
      "1": {
        amount: 2,
        category: "holdBag",
        restrictions: {
          weight: 15,
          height: 52,
          width: 26,
          length: 78,
          dimensions_sum: 156,
        },
      },
      "2": {
        amount: 1,
        category: "holdBag",
        restrictions: {
          weight: 25,
          height: 52,
          width: 26,
          length: 78,
          dimensions_sum: 156,
        },
        conditions: {
          is_priority: ["FR", "W6"],
        },
      },
    },
  },
  {
    originalIndex: 10,
    bagType: "holdBag",
    price: {
      currency: "EUR",
      amount: 60,
    },
    items: {
      "1": {
        amount: 2,
        category: "holdBag",
        restrictions: {
          weight: 15,
          height: 52,
          width: 26,
          length: 78,
          dimensions_sum: 156,
        },
      },
      "2": {
        amount: 2,
        category: "holdBag",
        restrictions: {
          weight: 25,
          height: 52,
          width: 26,
          length: 78,
          dimensions_sum: 156,
        },
        conditions: {
          is_priority: ["FR", "W6"],
        },
      },
    },
  },
];

export default data;
