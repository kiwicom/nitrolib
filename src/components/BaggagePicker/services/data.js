// @flow
export const baggageData = {
  definitions: {
    handBag: [
      {
        conditions: {
          passengerGroups: ["adult"],
        },
        price: {
          currency: "EUR",
          amount: 0,
          base: 0,
          merchant: null,
          service: 0,
          serviceFlat: 0,
        },
        category: "personalItem",
        restrictions: {
          weight: 5,
          height: 20,
          width: 20,
          length: 20,
          dimensionsSum: null,
        },
      },
      {
        conditions: {
          passengerGroups: ["adult"],
        },
        price: {
          currency: "EUR",
          amount: 10,
          base: 0,
          merchant: null,
          service: 0,
          serviceFlat: 0,
        },
        category: "cabinBag",
        restrictions: {
          weight: 10,
          height: 35,
          width: 20,
          length: 45,
          dimensionsSum: null,
        },
      },
    ],
    holdBag: [
      {
        conditions: {
          passengerGroups: ["adult"],
        },
        price: {
          currency: "EUR",
          amount: 10,
          base: 0,
          merchant: null,
          service: 0,
          serviceFlat: 0,
        },
        category: "holdBag",
        restrictions: {
          weight: 10,
          height: 52,
          width: 26,
          length: 78,
          dimensionsSum: 156,
        },
      },
      {
        conditions: {
          passengerGroups: ["adult"],
        },
        price: {
          currency: "EUR",
          amount: 20,
          base: 20,
          merchant: null,
          service: 5,
          serviceFlat: 0,
        },
        category: "holdBag",
        restrictions: {
          weight: 15,
          height: 52,
          width: 26,
          length: 78,
          dimensionsSum: 156,
        },
      },
      {
        conditions: {
          passengerGroups: ["adult"],
          isPriority: ["FR", "W6"],
        },
        price: {
          currency: "EUR",
          amount: 30,
          base: 0,
          merchant: null,
          service: 0,
          serviceFlat: 0,
        },
        category: "holdBag",
        restrictions: {
          weight: 25,
          height: 52,
          width: 26,
          length: 78,
          dimensionsSum: 156,
        },
      },
    ],
  },
  combinations: {
    handBag: [
      {
        conditions: {
          passengerGroups: ["adult"],
        },
        indices: [0],
        price: {
          currency: "EUR",
          amount: 0,
          base: 0,
          merchant: null,
          service: 0,
          serviceFlat: 0,
        },
      },
      {
        conditions: {
          passengerGroups: ["adult"],
        },
        indices: [1],
        price: {
          currency: "EUR",
          amount: 10,
          base: 0,
          merchant: null,
          service: 0,
          serviceFlat: 0,
        },
      },
      {
        conditions: {
          passengerGroups: ["adult"],
        },
        indices: [0, 1],
        price: {
          currency: "EUR",
          amount: 10,
          base: 0,
          merchant: null,
          service: 0,
          serviceFlat: 0,
        },
      },
      {
        conditions: {
          passengerGroups: ["adult"],
        },
        indices: [0, 1, 1],
        price: {
          currency: "EUR",
          amount: 20,
          base: 0,
          merchant: null,
          service: 0,
          serviceFlat: 0,
        },
      },
      {
        conditions: {
          passengerGroups: ["adult"],
        },
        indices: [0, 1, 1, 1],
        price: {
          currency: "EUR",
          amount: 30,
          base: 0,
          merchant: null,
          service: 0,
          serviceFlat: 0,
        },
      },
    ],
    holdBag: [
      {
        conditions: {
          passengerGroups: ["adult"],
        },
        indices: [],
        price: {
          currency: "EUR",
          amount: 0,
          base: 0,
          merchant: null,
          service: 0,
          serviceFlat: 0,
        },
      },
      {
        conditions: {
          passengerGroups: ["adult"],
        },
        indices: [0],
        price: {
          currency: "EUR",
          amount: 10,
          base: 0,
          merchant: null,
          service: 0,
          serviceFlat: 0,
        },
      },
      {
        conditions: {
          passengerGroups: ["adult"],
        },
        indices: [1],
        price: {
          currency: "EUR",
          amount: 20,
          base: 0,
          merchant: null,
          service: 0,
          serviceFlat: 0,
        },
      },
      {
        conditions: {
          passengerGroups: ["adult"],
        },
        indices: [2],
        price: {
          currency: "EUR",
          amount: 30,
          base: 0,
          merchant: null,
          service: 0,
          serviceFlat: 0,
        },
      },
      {
        conditions: {
          passengerGroups: ["adult"],
        },
        indices: [1, 1, 2],
        price: {
          currency: "EUR",
          amount: 70,
          base: 0,
          merchant: null,
          service: 0,
          serviceFlat: 0,
        },
      },
      {
        conditions: {
          passengerGroups: ["adult"],
        },
        indices: [1, 1, 2, 2],
        price: {
          currency: "EUR",
          amount: 100,
          base: 0,
          merchant: null,
          service: 0,
          serviceFlat: 0,
        },
      },
    ],
  },
};

export const emptyData = {
  definitions: {
    handBag: [],
    holdBag: [],
  },
  combinations: {
    handBag: [],
    holdBag: [],
  },
};
