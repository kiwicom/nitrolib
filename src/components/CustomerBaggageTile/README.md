# CustomerBaggageTile

- renders baggage tile for customer
- component should be used in baggage ordering process
- component has features:
  - order statuses
  - differ newly selected baggages from current

**Example:**

```js
<CustomerBaggageTile
  firstName="Vaclav"
  lastName="Havel"
  gender="male"
  isProcessing={false},
  current={{
    handBag: 1,
    holdBag: 1,
  }}
  selected={{
    handBag: 2,
    holdBag: 2,
  }}
  onClick={ () => {}}
  baggage={baggageData}
/>
```
