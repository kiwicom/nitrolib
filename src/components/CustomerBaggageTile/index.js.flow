// @flow strict
import type {
  BaggageType,
  Gender,
  HoldBagTileDefinition,
  HandBagTileDefinition,
} from "../../records/Baggage";

type Props = {|
  firstName: string,
  middleName?: string,
  lastName: string,
  gender: Gender,
  dayOfBirth?: string,
  isProcessing: boolean,
  current?: {
    handBag: number,
    holdBag: number,
  },
  selected?: {
    handBag: number,
    holdBag: number,
  },
  newDefinitions?: (HoldBagTileDefinition | HandBagTileDefinition)[],
  onClick?: () => void,
  baggage: BaggageType,
|};

declare export default React.ComponentType<Props>;
