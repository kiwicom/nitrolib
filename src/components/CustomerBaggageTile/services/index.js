// @flow strict
import R from "ramda";
import { Decimal } from "decimal.js";

import { getTotalPrice } from "../../../services/baggage/utils";
import type {
  HoldBagTileDefinition,
  BaggageCategory,
  HandBagTileDefinition,
  Combinations,
  OrderStatusType,
  BaggageType,
} from "../../../records/Baggage";

export const handleIsCurrentFlag = (
  definitions: (HoldBagTileDefinition | HandBagTileDefinition)[],
  commonDefIndices: number[],
) => {
  const memo = [...commonDefIndices];
  return definitions.reduce((acc, def) => {
    if (memo.some(i => i === def.originalIndex)) {
      def.isCurrent = true; // eslint-disable-line
      memo.splice(memo.findIndex(i => i === def.originalIndex), 1); // eslint-disable-line
    }
    // $FlowExpected: Union types issue
    return acc.concat(def);
  }, []);
};

export const filterNewDefinitions = (
  definitions?: (HoldBagTileDefinition | HandBagTileDefinition)[],
  bagType: BaggageCategory,
): (HoldBagTileDefinition | HandBagTileDefinition)[] => {
  if (definitions) {
    return definitions.filter(def => {
      if (bagType === "handBag") {
        return def.category === "cabinBag" || def.category === "personalItem";
      }
      return def.category === "holdBag";
    });
  }
  return [];
};
type GetDefinitionsArgs = {
  current?: {
    handBag: number, // index of combination
    holdBag: number, // index of combination
  },
  selected?: {
    handBag: number, // index of combination
    holdBag: number, // index of combination
  },
  baggage: BaggageType,
  bagType: BaggageCategory,
  newDefinitions?: (HoldBagTileDefinition | HandBagTileDefinition)[],
};
export const getDefinitions = ({
  current,
  selected,
  baggage,
  bagType,
  newDefinitions,
}: GetDefinitionsArgs): (HoldBagTileDefinition | HandBagTileDefinition)[] => {
  const { definitions, combinations } = baggage;
  const currentCombination = current && current[bagType];
  const selectedCombination = selected && selected[bagType];

  if (typeof selectedCombination === "number" && typeof currentCombination === "number") {
    const currentIndices = combinations[bagType][currentCombination].indices;
    const selectedIndices = combinations[bagType][selectedCombination].indices;
    const newDefinitionsIndices = R.intersection(selectedIndices, currentIndices);
    const selectedDef = selectedIndices.map(index => {
      // $FlowIssue: https://github.com/facebook/flow/issues/2892
      return {
        originalIndex: index,
        isCurrent: false,
        ...definitions[bagType][index],
      };
    });

    return handleIsCurrentFlag(selectedDef, newDefinitionsIndices);
  }
  return filterNewDefinitions(newDefinitions, bagType);
};

type CalculatePriceArgs = {
  current?: {
    handBag: number, // index of combination
    holdBag: number, // index of combination
  },
  selected?: {
    handBag: number, // index of combination
    holdBag: number, // index of combination
  },
  combinations: Combinations,
};
export const calculatePrice = ({
  selected,
  current,
  combinations,
}: CalculatePriceArgs): number | null => {
  if (selected && current) {
    const selectedPrice = getTotalPrice({
      combinationIndices: { handBag: [selected.handBag], holdBag: [selected.holdBag] },
      combinations,
    });
    const currentPrice = getTotalPrice({
      combinationIndices: { handBag: [current.handBag], holdBag: [current.holdBag] },
      combinations,
    });
    const diff = new Decimal(selectedPrice).minus(currentPrice).toFixed(2);
    return new Decimal(diff).toNumber();
  }
  return null;
};

type GetStatusArgs = {
  current?: {
    handBag: number, // index of combination
    holdBag: number, // index of combination
  },
  selected?: {
    handBag: number, // index of combination
    holdBag: number, // index of combination
  },
  isProcessing: boolean,
};
export const getStatus = ({
  current,
  selected,
  isProcessing,
}: GetStatusArgs): OrderStatusType | null => {
  if (current && selected && R.equals(current, selected)) {
    return null;
  }
  if (isProcessing) {
    return "processing";
  }
  if (current && selected) {
    return "unpaid";
  }
  return "notAvailable";
};
