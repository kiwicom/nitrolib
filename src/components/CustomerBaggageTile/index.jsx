// @flow strict
import * as React from "react";
import styled from "styled-components";
import Tile from "@kiwicom/orbit-components/lib/Tile";

import Title from "./components/Title";
import Content from "./components/Content";
import { getDefinitions, calculatePrice, getStatus } from "./services/index";
import type {
  BaggageType,
  Gender,
  HoldBagTileDefinition,
  HandBagTileDefinition,
} from "../../records/Baggage";

const Wrapper = styled.div`
  div:last-child > svg {
    display: ${({ isIconShown }: { isIconShown: boolean }) => (isIconShown ? "inline" : "none")};
  }
`;

type Props = {
  firstName: string,
  middleName?: string,
  lastName: string,
  gender: Gender,
  dayOfBirth?: string,
  isProcessing: boolean,
  current?: {
    handBag: number, // index of combination
    holdBag: number, // index of combination
  },
  selected?: {
    handBag: number, // index of combination
    holdBag: number, // index of combination
  },
  newDefinitions?: (HoldBagTileDefinition | HandBagTileDefinition)[],
  onClick?: () => void,
  baggage: BaggageType,
};

const CustomerBaggageTile = ({
  firstName,
  middleName,
  lastName,
  gender,
  dayOfBirth,
  isProcessing,
  current,
  selected,
  newDefinitions,
  onClick,
  baggage,
}: Props) => {
  const { combinations } = baggage;

  const handBag = getDefinitions({
    current,
    selected,
    baggage,
    newDefinitions,
    bagType: "handBag",
  });
  const holdBag = getDefinitions({
    current,
    selected,
    baggage,
    newDefinitions,
    bagType: "holdBag",
  });
  const status = getStatus({ current, selected, isProcessing });
  return (
    <Wrapper isIconShown={!!onClick}>
      <Tile
        dataTest={`CustomerBaggageTile-${status || "none"}`}
        onClick={onClick && onClick}
        title={
          <Title
            gender={gender}
            firstName={firstName}
            middleName={middleName}
            lastName={lastName}
            orderStatus={status}
            price={calculatePrice({ selected, current, combinations })}
            dayOfBirth={dayOfBirth}
          />
        }
        description={<Content definitions={[...handBag, ...holdBag]} orderStatus={status} />}
      />
    </Wrapper>
  );
};

export default CustomerBaggageTile;
