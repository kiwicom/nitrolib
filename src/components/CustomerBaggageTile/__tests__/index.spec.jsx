// @flow strict
import * as React from "react";
import { shallow } from "enzyme";

import CustomerBaggageTile from "../index";
import data from "../services/data";

const props = {
  firstName: "Oliver",
  middleName: undefined,
  lastName: "Dlouhy",
  gender: "male",
  dayOfBirth: undefined,
  isProcessing: false,
  current: { handBag: 1, holdBag: 1 },
  selected: { handBag: 3, holdBag: 4 },
  definitions: undefined,
  onClick: () => console.log("clicked"), // eslint-disable-line
  baggage: data,
};

describe("#CustomerBaggageTile", () => {
  test("renders", () => {
    const wrapper = shallow(<CustomerBaggageTile {...props} />);
    expect(wrapper).toMatchSnapshot();
  });
});
