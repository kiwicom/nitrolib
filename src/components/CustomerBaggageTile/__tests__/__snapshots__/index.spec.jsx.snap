// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#CustomerBaggageTile renders 1`] = `
<Tile
  description={
    <TileContent
      handBags={
        Array [
          Object {
            "category": "personalItem",
            "restrictions": Object {
              "dimensionsSum": null,
              "height": 20,
              "length": 20,
              "weight": 5,
              "width": 20,
            },
          },
          Object {
            "category": "cabinBag",
            "restrictions": Object {
              "dimensionsSum": null,
              "height": 35,
              "length": 45,
              "weight": 10,
              "width": 20,
            },
          },
          Object {
            "category": "cabinBag",
            "restrictions": Object {
              "dimensionsSum": null,
              "height": 35,
              "length": 45,
              "weight": 10,
              "width": 20,
            },
          },
        ]
      }
      holdBags={
        Array [
          Object {
            "category": "holdBag",
            "restrictions": Object {
              "dimensionsSum": 156,
              "height": 52,
              "length": 78,
              "weight": 15,
              "width": 26,
            },
          },
          Object {
            "category": "holdBag",
            "restrictions": Object {
              "dimensionsSum": 156,
              "height": 52,
              "length": 78,
              "weight": 15,
              "width": 26,
            },
          },
          Object {
            "category": "holdBag",
            "restrictions": Object {
              "dimensionsSum": 156,
              "height": 52,
              "length": 78,
              "weight": 25,
              "width": 26,
            },
          },
        ]
      }
      orderStatus="unpaid"
    />
  }
  onClick={[Function]}
  title={
    <Title
      firstName="Oliver"
      gender="male"
      lastName="Dlouhy"
      orderStatus="unpaid"
      price={
        Object {
          "amount": 21,
          "base": 21,
          "currency": "EUR",
          "service": 0,
          "serviceFlat": 0,
        }
      }
    />
  }
/>
`;
